p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/FtpTransportProtocolHandler.java

```diff
@@ -0,0 +1,217 @@
+        final FTPClient client = getClient(uri);
+
+        if (!client.hasFeature(FTPCmd.MDTM)) {
+            logger.debug("Could not retrieve the last modification timestamp for: " + uri);
+            return -1;
+        }
+
+        final FTPFile file = client.mdtmFile(uri.getPath());
+
+        if (file == null) {
+            throw new FileNotFoundException("Could not find file: " + uri);
+        }
+
+        return file.getTimestampInstant().toEpochMilli();
+    }
+
+    @Override
+    public File getFile(final URI uri) throws IOException {
+        final File localFile = getLocalFile(uri);
+
+        if (cacheConfig.isOffline()) {
+            if (!localFile.isFile()) {
+                throw new IOException("Maven is offline and the requested file does not exist locally: " + uri);
+            }
+
+            return localFile;
+        }
+
+        final FTPClient client = getClient(uri);
+        final String remotePath = uri.getPath();
+        final FTPFile remoteFile = getRemoteFile(client, remotePath);
+        final boolean isRemoteMissing = remoteFile == null;
+
+        if (localFile.isFile() && (isRemoteMissing || !mustRefresh(localFile, remoteFile))) {
+            return localFile;
+        }
+
+        if (isRemoteMissing) {
+            throw new FileNotFoundException("Could not find file: " + uri);
+        }
+
+        final File parent = FileUtils.createParentDirectories(localFile);
+        final File tempFile = Files.createTempFile(parent.toPath(), "download", ".tmp").toFile();
+        tempFile.deleteOnExit();
+
+        try (final OutputStream os = new FileOutputStream(tempFile)) {
+            if (!client.retrieveFile(remotePath, os)) {
+                final String message = client.getReplyString();
+                throw new IOException(String.format("Error retrieving file: %s. Message: %s", remotePath, message));
+            }
+        } catch (final IOException e) {
+            tempFile.delete();
+            throw e;
+        }
+
+        if (localFile.isFile()) {
+            FileUtils.forceDelete(localFile);
+        }
+
+        FileUtils.moveFile(tempFile, localFile);
+        localFile.setLastModified(remoteFile.getTimestampInstant().toEpochMilli());
+        return localFile;
+    }
+
+    public void dispose() {
+        for (final Map.Entry<String, FTPClient> entry : clients.entrySet()) {
+            final FTPClient client = entry.getValue();
+
+            try {
+                // We check if the connection is still active, as it might
+                // have been dropped earlier because of a timeout
+                if (client.isAvailable()) {
+                    client.logout();
+                    client.disconnect();
+                }
+            } catch (final FTPConnectionClosedException e) {
+                // Connection already closed by the host
+            } catch (final IOException e) {
+                logger.error("Error disconnecting from host: " + entry.getKey(), e);
```

Comment:
I think debug is enough, as there is not much to do anyways...

Reply:
I always prefer to see the actual error + stacktrace. Sometimes FTP servers have issues and errors can be of help.
Timeouts are not shown, so other errors might be of relevance.

Reply:
You can always enable -X if it is of any interest, but errors when closing a connection only pollute the usual logs.

Reply:
@laeubi not convinced by that, mostly because debug logging is rarely on, so you'd never know in the first place, plus it's a single stack trace.
What I might do is display a message only as error, and ask for debug to display the full stack trace, sounds reasonable?

Reply:
Not really, when closing fails, and the maven process end, there is literally no need to scare the user, what should happen here? The process will exit and any socket will forcefully be closed anyways, so there is nothing to worry about.

Reply:
@laeubi the socket is closed but as a user I want to know if there has been an issue anyway. From a tool perspective it's important the user is always aware of errors, even if the message is concise it can be expanded with debug.

Reply:
I can downgrade to warning, that's acceptable

Reply:
But what error? From user POV there is no error, the build has succeed and we are just cleaning up resources, so this "problem" will never influence the build outcome and thus must not be shown if not debug is requested.

Reply:
Well imho a connection error should be notified anyway, as warning at least. FTP repos are going to be used mostly internally and the people that manage the build might manage the server too (I do devops and I manage the full stack), and might be interested in being notified of connection issues. Depending on the type of error the client or sever might be tweaked or at least the server log can be inspected. If we show that only in debug mode it's never going to be seen, or very rarely

Reply:
If no one notice a problem, then there is no problem, Tycho is not an FTP-Server-Debugging-Tool.